### ISSUE TITLE
Rails 7: ActiveStorage setup migration fails to correctly dump the new schema.rb (with primary key type: UUID)

### ISSUE THREAD

**Issue by john-999** [rails/rails#49752]
### Steps to reproduce

After running the [ActiveStorage setup](https://edgeguides.rubyonrails.org/active_storage_overview.html#setup), the migration producing the new `schema.rb` dump appears to fail, the beginning of it looking like this:

```ruby
ActiveRecord::Schema[7.0].define(version: 2023_09_29_141433) do
# Could not dump table "active_storage_attachments" because of following StandardError
#   Unknown type 'uuid' for column 'id'

# Could not dump table "active_storage_blobs" because of following StandardError
#   Unknown type 'uuid' for column 'id'

# Could not dump table "active_storage_variant_records" because of following StandardError
#   Unknown type 'uuid' for column 'id'
...
```

In an initializer file, I have set:
```ruby
Rails.application.config.generators { |g| g.orm :active_record, primary_key_type: :uuid }
```

However, looking at the generated DB tables in phpMyAdmin, all key and foreign-key values in the "Type" columns correctly show "uuid".

### System configuration
**Rails version**: `7.0.4.3`

**Ruby version**: `3.0.1p64` (2021-04-05 revision 0fb782ee38) [x86_64-linux]

**Comment by fatkodima** [comment]
I cannot reproduce. Could you create a sample rails app that reproduces the issue and share with us? Also note that Rails 7.0 is not supported for bug fixed (see https://guides.rubyonrails.org/maintenance_policy.html), so test against 7.1+.

**Comment by john-999** [comment]
Yes, I have reproduced the issue in a new sample `Rails 7.1.1` app: [as_tables_in_schema_dump.zip](https://github.com/rails/rails/files/13168761/as_tables_in_schema_dump.zip)

1. Download the attached .zipped app folder, and uncompress it to your preferred location.
2. Find the error messages in the file `db/schema.rb`.

If you want to work on/with this app directly:

3. Make sure the above mentioned `System configuration` is satisfied.
4. Ensure presence of gems: `$ bundle install`
5. Set your database user and password: `$ rails credentials:edit`
6. Create the new database: `$ rails db:create`
7. Create the ActiveStorage tables: `$ rails db:migrate`
8. Optionally: Reproduce the error by regenerating the file `db/schema.rb`: `$ rails db:rollback`, then `$ rails db:migrate`, see the error messages in the file `db/schema.rb`.

If you wish to recreate your own sample app to work with, here are the steps:
1. Make sure the above mentioned `System configuration` is satisfied.
2. Create the new app: `$ rails new as_tables_in_schema_dump`
3. Configure + install gem in `Gemfile`: `gem "mysql2", "~> 0.5.3"`, then `$ bundle install`.
4. Configure file: `database.yml`
5. Set your database user and password: `$ rails credentials:edit`
6. Create the new database: `$ rails db:create`
7. Add in file: `config/application.rb`:
```
config.generators do |g|
  g.orm :active_record, primary_key_type: :uuid
end
```
8. Set up ActiveStorage: `$ rails active_storage:install`
9. Run the copied migrations: `$ rails db:migrate`
10. See the errors in the file `db/schema.rb`.

**Comment by fatkodima** [comment]
Thanks, was able to reproduce. Will look into it.

**Comment by fatkodima** [comment]
Was this working before? Seems like tables with `uuid` columns are not dumped in `schema.rb`. I haven't investigated, but does Rails support uuids for MySQL at all?

cc @matthewd @kamipo @rafaelfranca

**Comment by john-999** [comment]
I don't know if `uuid` columns in general have worked before in Rails, for MariaDB/MySQL.

FWIW, what I do know is that `uuid` is an important feature, that MariaDB is a popular DB, and that Rails correctly creates `uuid` columns in MariaDB which has support for `uuid`.

(It is only the `schema.rb` dump in Rails that fails).

**Comment by rafaelfranca** [comment]
If it is supported in the mysql adapter, it should also work for MariaDB

**Comment by john-999** [comment]
@rafaelfranca - If you're referring to the `gem mysql2`, then take a look at what the creator of the gem says: https://github.com/brianmario/mysql2/issues/1337

So, if I understand correctly, Rails `uuid` support for MariaDB/MySQL entirely depends on the intention of @brianmario, correct?

**Comment by fatkodima** [comment]
`uuid`s are working in migrations, because we can specify basically any type in `t.column :column_name, :some_type` and it will work. But it is not dumped into the schema file.

I was asking if it was supported, because I was surprised to not found it in mysql rails adapter, but it is explicitly supported in postgres adapter. Maybe I am missing something?

I can try to work on introducing it into rails, if desired.

**Comment by rafaelfranca** [comment]
Yes. We should support. I'd consider that a bug.

**Comment by fatkodima** [comment]
Ok, I will try to come up with something on the weekend.

**Comment by rafaelfranca** [comment]
> So, if I understand correctly, Rails uuid support for MariaDB/MySQL entirely depends on the intention of @brianmario, correct?

No, I mean, if it works when you use the `adapter: mysql2` with MySQL and not with MariaDB, we should support when you use with MariaDB as well.

If it isn't support with neither of the databases, we should add support regardless, but I was trying to understand if it is MariaDB specific, or not.

**Comment by fatkodima** [comment]
It does not work neither with MySQL nor with MariaDB. 
Fortunately, there is a native `uuid` type in MariaDB, but there is no one for MySQL. MySQL has a `UUID()` function, but for actually storing it in the table people use different hacks on top of columns defined like `varchar(36)` or `binary(16)` to mimic uuids.

**Comment by rafaelfranca** [comment]
Got it, so this would be a new feature. That is fine, let's build it, but it will only be present in 7.2.

**Comment by zgid123** [comment]
for my case, it is `multilinestring` and `linestring`

```rb
#   Unknown type 'linestring' for column 'schedule'
```
